singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,420 @@
+                    "host=" + host, "stage=" + stage.toString(),
+                    "logName=" + event.getLoggingAuditHeaders().getLogName());
+            eventTriedCount.remove(event.getLoggingAuditHeaders());
+          }
+        }
+      } catch (InterruptedException e) {
+        LOG.warn("[{}] got interrupted when polling the queue and while loop is ended!",
+            Thread.currentThread().getName(), e);
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_DEQUEUE_INTERRUPTED_EXCEPTION, 1,
+                "host=" + host, "stage=" + stage.toString());
+        break;
+      } catch (Exception e) {
+        LOG.warn("Exit the while loop and finish the thread execution due to exception: ", e);
+        OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_EXCEPTION, 1,
+            "host=" + host, "stage=" + stage.toString());
+        break;
+      }
+    }
+  }
+
+
+  public class KafkaProducerCallback implements Callback {
+
+    private LoggingAuditEvent event;
+
+    public KafkaProducerCallback(LoggingAuditEvent event) {
+      this.event = event;
+    }
+
+    public void checkAndEnqueueWhenSendFailed(RecordMetadata recordMetadata, Exception e) {
+      // if exception thrown (i.e. the send failed), the partition is added to badPartitions.
+      badPartitions.add(recordMetadata.partition());
+      OpenTsdbMetricConverter
+          .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_PARTITION_ERROR, 1,
+              "host=" + host, "stage=" + stage.toString(), "topic=" + topic,
+              "partition=" + recordMetadata.partition());
+
+      // retry the failed event by inserting it at the beginning of the deque.
+      // If number of tries reaches 3, meaning that 3 partitions have been tried sending to but
+      // still failed, this event is dropped.
+
+      Integer count = eventTriedCount.get(event.getLoggingAuditHeaders());
+      if (count == null){
+        eventTriedCount.put(event.getLoggingAuditHeaders(), 1);
+        OpenTsdbMetricConverter
+            .gauge(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_EVENTS_RETRIED,
+                eventTriedCount.size(), "host=" + host, "stage=" + stage.toString(),
+                "topic=" + topic);
+      } else {
+        if (count >= NUM_OF_PARTITIONS_TO_TRY_SENDING) {
+          LOG.debug("Failed to send audit event after trying {} partitions. Drop this event.",
+              count);
+          OpenTsdbMetricConverter
+              .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_EVENTS_DROPPED, 1,
+                  "host=" + host, "stage=" + stage.toString(),
+                  "logName=" + event.getLoggingAuditHeaders().getLogName());
+          eventTriedCount.remove(event.getLoggingAuditHeaders());
+        } else {
+          eventTriedCount.put(event.getLoggingAuditHeaders(), count + 1);
+          boolean success = false;
```

Comment:
remove line 306 and update line 308 as 

    boolean success = queue.offerFirst(event, 3, TimeUnit.SECONDS);

Reply:
made the change in the update of the PR.